src/cudamatrix/cu-sparse-matrix.cc

```diff
@@ -42,6 +42,205 @@
 namespace kaldi {
 
 template <typename Real>
+RowElement<Real>* CuRowSparseMatrix<Real>::Data() {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    if (data_.Dim() == 0)
+      return NULL;
+    else
+      return data_.Data();
+  } else
+#endif
+  {
+    return NULL;
+  }
+}
+
+template <typename Real>
+const RowElement<Real>* CuRowSparseMatrix<Real>::Data() const {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    if (data_.Dim() == 0)
+      return NULL;
+    else
+      return data_.Data();
+  } else
+#endif
+  {
+    return NULL;
+  }
+}
+
+template <typename Real>
+CuRowSparseMatrix<Real>& CuRowSparseMatrix<Real>::operator = (
+    const SparseMatrix<Real> &smat) {
+  this->CopyFromSmat(smat);
+  return *this;
+}
+
+template <typename Real>
+CuRowSparseMatrix<Real>& CuRowSparseMatrix<Real>::operator = (
+    const CuRowSparseMatrix<Real> &smat) {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    data_ = smat.data_;
+    num_rows_ = smat.num_rows_;
+    num_cols_ = smat.num_cols_;
+    elements_per_row_ = smat.elements_per_row_;
+  } else
+#endif
+  {
+    this->Mat() = smat.Mat();
+  }
+  return *this;
+}
+
+template <typename Real>
+void CuRowSparseMatrix<Real>::Swap(SparseMatrix<Real> *smat) {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    CuRowSparseMatrix<Real> tmp(*smat);
+    Swap(&tmp);
+    tmp.CopyToSmat(smat);
+  } else
+#endif
+  {
+    num_rows_ = smat->NumRows();
+    num_cols_ = smat->NumCols();
+    Mat().Swap(smat);
+  }
+}
+
+template <typename Real>
+void CuRowSparseMatrix<Real>::Swap(CuRowSparseMatrix<Real> *smat) {
+#if HAVE_CUDA == 1
```

Comment:
This Swap() implementation is more complicated than it needs to be.
Just do a shallow swap of each element, i.e.:
cpu_rows_.swap(other->cpu_rows);
std::swap(num_rows_, other->num_rows_);
std::swap(num_cols_, other->num_cols_);
std::swap(elements_per_row_, other->elements_per_row_);
data_.Swap(&(other->data_));

Even if a data-member isn't being used, swapping emepty members is OK.